rules/coding-style/src/Rector/String_/UseClassKeywordForClassNameResolutionRector.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\CodingStyle\Rector\String_;
+
+use Nette\Utils\Strings;
+use PhpParser\Node;
+use PhpParser\Node\Expr\BinaryOp\Concat;
+use PhpParser\Node\Expr\ClassConstFetch;
+use PhpParser\Node\Name\FullyQualified;
+use PhpParser\Node\Scalar\String_;
+use Rector\Core\Rector\AbstractRector;
+use Rector\Core\RectorDefinition\CodeSample;
+use Rector\Core\RectorDefinition\RectorDefinition;
+
+/**
+ * @see \Rector\CodingStyle\Tests\Rector\String_\UseClassKeywordForClassNameResolutionRector\UseClassKeywordForClassNameResolutionTest
+ */
+final class UseClassKeywordForClassNameResolutionRector extends AbstractRector
+{
+    public function getDefinition(): RectorDefinition
+    {
+        return new RectorDefinition(
+            'Use `class` keyword for class name resolution in string instead of hardcoded string reference',
+            [
+                new CodeSample(
+                    <<<'PHP'
+$this->write(
+                '<?php if (App\SomeClass::createCache($netteCacheStorage, %var,  $this->global->cacheStack, %node.array?)) { ?>',
+                md5('md5')
+            );
+PHP
+                    ,
+                    <<<'PHP'
+$this->write(
+                '<?php if (' . \App\SomeClass::class . '::createCache($netteCacheStorage, %var,  $this->global->cacheStack, %node.array?)) { ?>',
+                \md5('md5')
+            );
+PHP
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [String_::class];
+    }
+
+    /**
+     * @param String_ $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if (substr_count($node->value, '::') !== 1) {
+            return null;
+        }
+        // a possible static call reference
+        [$before, $possibleClass, $after] = Strings::split($node->value, '#([\\\\a-zA-Z0-9_\\x80-\\xff]*)::#');
+
+        if (! $possibleClass || ! class_exists($possibleClass)) {
+            return null;
+        }
+
+        $classConstFetch = new ClassConstFetch(new FullyQualified(ltrim($possibleClass, '\\')), 'class');
+
+        $concat = new Concat($classConstFetch, new String_('::' . $after));
+        if (! empty($before)) {
```

Comment:
Not sure what `$before` is

Reply:
Before is everything before the class name. See the regex https://regex101.com/r/Vv41Qr/1
In some cases, it might be empty and I don't think we would like to generate empty string before the class name.
```diff
-'' . ClassName . '::method()'
+ClassName . '::method()'
```

Reply:
I see. Please add one test case for that too.

Reply:
Ready